flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java

```diff
@@ -877,4 +899,610 @@ private String getSensorPositionInfo(Asset asset) {
+      Map<String, Map<String, String>> idMaps = new HashMap<>();
+      Map<String, String> metricFormula = asset.getMetricsformulars();
+      if(metricFormula == null || metricFormula.get(FlowgateConstant.SENSOR) == null) {
+         return idMaps;
+      }
+      String sensorFormulasInfo = metricFormula.get(FlowgateConstant.SENSOR);
+      Map<String, Map<String, String>> sensorFormulars =
+            asset.metricsFormulaToMap(sensorFormulasInfo, new TypeReference<Map<String, Map<String, String>>>() {});
+      Map<String, String> idAndMetricNameMap = new HashMap<String, String>();
+      Map<String, String> idAndLocationMap = new HashMap<String, String>();
+      //generate the map of sensorId and metricName
+      //generate the map of sensorId and location
+      for (Map.Entry<String, Map<String, String>> sensorFormula : sensorFormulars.entrySet()) {
+         Map<String, String> locationAndIdMap = sensorFormula.getValue();
+         String metricName = sensorFormula.getKey();
+         for(Map.Entry<String, String> locationAndIdEntry : locationAndIdMap.entrySet()) {
+            String formula = locationAndIdEntry.getValue();
+            String location = locationAndIdEntry.getKey();
+            String ids[] = formula.split("\\+|-|\\*|/|\\(|\\)");
+            for(String sensorAssetId : ids) {
+               idAndMetricNameMap.put(sensorAssetId, metricName);
+               idAndLocationMap.put(sensorAssetId, location);
+            }
+         }
+      }
+      if(idAndMetricNameMap.isEmpty()) {
+         return idMaps;
+      }
+      idMaps.put(IDNAMEMAP, idAndMetricNameMap);
+      idMaps.put(IDLOCATIONMAP, idAndLocationMap);
+      return idMaps;
+   }
+
+   private List<MetricData> generateSensorsMetricDataForPDU(String location, List<ValueUnit> valueUnits){
+      List<MetricData> result = new ArrayList<MetricData>();
+      if(location.indexOf(FlowgateConstant.SEPARATOR) > -1) {
+         location = location.replace(FlowgateConstant.SEPARATOR, FlowgateConstant.UNDERLINE);
+      }
+      for (ValueUnit valueunit : valueUnits) {
+         MetricData metricData = new MetricData();
+         metricData.setTimeStamp(valueunit.getTime());
+         metricData.setValue(valueunit.getValue());
+         metricData.setValueNum(valueunit.getValueNum());
+         String unit = databaseUnitAndOutputUnitMap.get(valueunit.getUnit());
+         if(unit == null) {
+            unit = valueunit.getUnit();
+         }
+         metricData.setUnit(unit);
+         switch (valueunit.getKey()) {
+         case MetricName.PDU_HUMIDITY:
+         case MetricName.PDU_TEMPERATURE:
+            metricData.setMetricName(String.format(pduValueunitKeyAndMetricNameMap.get(valueunit.getKey()),
+                  location));
+            result.add(metricData);
+            break;
+
+         default:
+            break;
+         }
+      }
+      return result;
+   }
+
+   private List<MetricData> generatePDUMetricDataForPDU(List<ValueUnit> valueUnits){
```

Comment:
convertPDUMetrics? not sure what is pdu metric for pdu?

Reply:
converPDUUseage